File: tests/python/test_elementary.py
"""Rigorous enclosure tests for the elementary functions, against a
high-precision mpmath oracle.

For each function and many deterministically-seeded input intervals, the true
mathematical function is sampled across the interval at ~200-digit precision;
every sample must lie within the computed interval result. This is the
authoritative check that the interval transcendentals are rigorous enclosures.
Structural/known-value checks live in the C++ suite.
"""

import math

import pytest

import pyintval as iv

mp = pytest.importorskip("mpmath")
mp.mp.prec = 220

I = iv.Interval
INF = math.inf


def _encloses(out: iv.Interval, value) -> bool:
    v = mp.mpf(value)
    lo = -mp.inf if out.lo == -INF else mp.mpf(out.lo)
    hi = mp.inf if out.hi == INF else mp.mpf(out.hi)
    return lo <= v <= hi


def _sample(a: float, b: float, n: int):
    # mpf arithmetic: k=0 -> exactly a, k=n-1 -> exactly b, all strictly inside.
    lo, hi = mp.mpf(a), mp.mpf(b)
    return [lo + (hi - lo) * mp.mpf(k) / (n - 1) for k in range(n)]


def _real_cbrt(t):
    return mp.sign(t) * mp.cbrt(abs(t))  # mp.cbrt gives the complex principal root


# (name, interval-fn, mpmath-fn, safe-domain-lo, safe-domain-hi)
UNARY = [
    ("exp", iv.exp, mp.exp, -700.0, 700.0),
    ("expm1", iv.expm1, mp.expm1, -700.0, 700.0),
    ("exp2", iv.exp2, lambda t: mp.power(2, t), -1000.0, 1000.0),
    ("exp10", iv.exp10, lambda t: mp.power(10, t), -300.0, 300.0),
    ("cbrt", iv.cbrt, _real_cbrt, -1e6, 1e6),
    ("sinh", iv.sinh, mp.sinh, -700.0, 700.0),
    ("cosh", iv.cosh, mp.cosh, -700.0, 700.0),
    ("tanh", iv.tanh, mp.tanh, -1e6, 1e6),
    ("asinh", iv.asinh, mp.asinh, -1e6, 1e6),
    ("atan", iv.atan, mp.atan, -1e6, 1e6),
    ("erf", iv.erf, mp.erf, -1e3, 1e3),
    ("erfc", iv.erfc, mp.erfc, -100.0, 100.0),
    ("sin", iv.sin, mp.sin, -1e3, 1e3),
    ("cos", iv.cos, mp.cos, -1e3, 1e3),
    ("tan", iv.tan, mp.tan, -1e3, 1e3),
    ("log", iv.log, mp.log, 1e-300, 1e300),
    ("log2", iv.log2, lambda t: mp.log(t, 2), 1e-300, 1e300),
    ("log10", iv.log10, mp.log10, 1e-300, 1e300),
    ("log1p", iv.log1p, mp.log1p, -0.9999, 1e300),
    ("asin", iv.asin, mp.asin, -1.0, 1.0),
    ("acos", iv.acos, mp.acos, -1.0, 1.0),
    ("atanh", iv.atanh, mp.atanh, -0.9999, 0.9999),
    ("acosh", iv.acosh, mp.acosh, 1.0, 1e6),
]


def _gen_interval(rng, dlo, dhi):
    """A random [lo, hi] strictly inside the safe domain [dlo, dhi]."""
    c = rng.uniform(dlo, dhi)
    span = dhi - dlo
    w = span * 10 ** rng.uniform(-12, -0.5)
    lo = max(c - w * rng.random(), dlo)
    hi = min(c + w * rng.random(), dhi)
    if lo > hi:
        lo, hi = hi, lo
    return lo, hi


@pytest.mark.parametrize("name,ivf,mpf,dlo,dhi", UNARY, ids=[u[0] for u in UNARY])
def test_unary_enclosure(name, ivf, mpf, dlo, dhi):
    import random

    rng = random.Random(1000 + UNARY.index((name, ivf, mpf, dlo, dhi)))
    for _ in range(500):
        lo, hi = _gen_interval(rng, dlo, dhi)
        out = ivf(I(lo, hi))
        assert not out.is_empty
        # tan asymptotes legitimately yield the entire line; nothing to enclose.
        if out.is_entire:
            continue
        for p in _sample(lo, hi, 15):
            assert _encloses(out, mpf(p)), f"{name}({float(p)!r}) not in {out}"


@pytest.mark.parametrize("name,ivf,mpf", [
    ("sin", iv.sin, mp.sin), ("cos", iv.cos, mp.cos),
], ids=["sin", "cos"])
def test_periodic_dense(name, ivf, mpf):
    import random

    rng = random.Random(0xC05 + (name == "cos"))
    for _ in range(2000):
        a = rng.uniform(-30, 30)
        b = a + 10 ** rng.uniform(-4, 1.2)
        out = ivf(I(a, b))
        for p in _sample(a, b, 60):  # dense: must catch interior peaks/troughs
            assert _encloses(out, mpf(p)), f"{name} peak missed on [{a},{b}] -> {out}"


def test_two_argument_enclosure():
    import random

    rng = random.Random(0xA7A2)
    for _ in range(3000):
        x1, x2 = sorted(rng.uniform(-30, 30) for _ in range(2))
        y1, y2 = sorted(rng.uniform(-30, 30) for _ in range(2))
        hy = iv.hypot(I(x1, x2), I(y1, y2))
        at = iv.atan2(I(y1, y2), I(x1, x2))
        for xx in _sample(x1, x2, 6):
            for yy in _sample(y1, y2, 6):
                assert _encloses(hy, mp.hypot(xx, yy))
                assert _encloses(at, mp.atan2(yy, xx))  # mpmath order is atan2(y, x)
        # pow with strictly positive base
        bx1, bx2 = sorted(rng.uniform(0.01, 20) for _ in range(2))
        pw = iv.pow(I(bx1, bx2), I(y1, y2))
        if pw.is_entire:
            continue
        for xx in _sample(bx1, bx2, 5):
            for yy in _sample(y1, y2, 5):
                assert _encloses(pw, mp.power(xx, yy))


def test_tightness_at_points():
    # A point interval must yield an enclosure only a couple of ulps wide.
    for f, val in [(iv.exp, 1.3), (iv.log, 3.7), (iv.sin, 0.9), (iv.cos, 2.1),
                   (iv.atan, 5.0), (iv.tanh, 0.6), (iv.cbrt, 9.0), (iv.erf, 0.4)]:
        out = f(I(val))
        assert out.wid <= 4 * math.ulp(max(abs(out.hi), abs(out.lo), 1e-300))


def test_known_identities():
    for x in [0.5, 1.0, 2.0, 10.0, 100.0]:
        assert x in iv.exp(iv.log(I(x)))
    for t in [0.3, 1.0, 2.5, 4.0]:
        s = iv.sin(I(t))
        c = iv.cos(I(t))
        assert 1.0 in (s * s + c * c)
    assert math.e in iv.e()
    assert math.pi in iv.pi()


def test_constants_are_tight():
    assert iv.pi().wid <= 4 * math.ulp(3.14159)
    assert iv.e().wid <= 4 * math.ulp(2.71828)

File: tests/cpp/test_elementary.cpp
// Structural and known-value tests for the interval elementary functions.
// The authoritative rigorous enclosure check (random inputs vs a high-precision
// mpmath oracle) lives in tests/python/test_elementary.py; this file pins the
// mathematical structure that must hold on every platform without an oracle:
// known values, domain handling, periodic extrema, unbounded limits, empty
// propagation, and tightness.

#include <cmath>
#include <limits>

#include "doctest/doctest.h"
#include "pyintval/elementary.hpp"

using namespace pyintval;

namespace {
constexpr double kInfv = std::numeric_limits<double>::infinity();
bool holds(double v, const Interval& x) { return is_member(v, x); }
}  // namespace

TEST_CASE("known values are enclosed") {
  CHECK(holds(1.0, exp(point(0.0))));
  CHECK(exp(point(1.0)).lo <= 2.718281828459045);
  CHECK(2.718281828459045 <= exp(point(1.0)).hi);
  CHECK(holds(0.0, log(point(1.0))));
  CHECK(holds(1.0, exp2(point(0.0))));
  CHECK(holds(8.0, exp2(point(3.0))));
  CHECK(holds(3.0, log2(point(8.0))));
  CHECK(holds(2.0, log10(point(100.0))));
  CHECK(holds(0.0, expm1(point(0.0))));
  CHECK(holds(0.0, log1p(point(0.0))));
  CHECK(holds(2.0, cbrt(point(8.0))));
  CHECK(holds(-3.0, cbrt(point(-27.0))));
  CHECK(holds(0.0, sin(point(0.0))));
  CHECK(holds(1.0, cos(point(0.0))));
  CHECK(holds(0.0, tan(point(0.0))));
  CHECK(holds(0.0, sinh(point(0.0))));
  CHECK(holds(1.0, cosh(point(0.0))));
  CHECK(holds(0.0, tanh(point(0.0))));
  CHECK(holds(0.0, asin(point(0.0))));
  CHECK(holds(0.0, atan(point(0.0))));
  CHECK(holds(0.0, asinh(point(0.0))));
  CHECK(holds(0.0, acosh(point(1.0))));
  CHECK(holds(0.0, atanh(point(0.0))));
  CHECK(holds(0.0, erf(point(0.0))));
  CHECK(holds(1.0, erfc(point(0.0))));
  CHECK(holds(5.0, hypot(point(3.0), point(4.0))));
}

TEST_CASE("bounded ranges are respected") {
  CHECK(subset(tanh(entire()), make(-1.0, 1.0)));
  CHECK(subset(erf(entire()), make(-1.0, 1.0)));
  CHECK(subset(atan(entire()), make(-2.0, 2.0)));  // within (-pi/2, pi/2)
  CHECK(subset(sin(make(-100.0, 100.0)), make(-1.0, 1.0)));
  CHECK(subset(cos(make(-100.0, 100.0)), make(-1.0, 1.0)));
  CHECK(cosh(make(-2.0, 3.0)).lo >= 1.0);
  CHECK(erfc(entire()).lo >= 0.0);
  CHECK(erfc(entire()).hi <= 2.0);
}

TEST_CASE("periodic extrema are captured") {
  // Interval spanning a full period -> full range.
  CHECK(equal(sin(make(0.0, 7.0)), make(-1.0, 1.0)));
  CHECK(equal(cos(make(0.0, 7.0)), make(-1.0, 1.0)));
  // Interval straddling pi/2 reaches the maximum 1.
  CHECK(sin(make(1.5, 1.7)).hi == 1.0);
  // Interval straddling pi reaches the minimum -1 for cos.
  CHECK(cos(make(3.0, 3.3)).lo == -1.0);
  // A narrow interval away from any extremum does NOT pin to +-1.
  CHECK(sin(make(0.1, 0.2)).hi < 1.0);
  CHECK(sin(make(0.1, 0.2)).lo > -1.0);
}

TEST_CASE("tan asymptotes give the entire line") {
  CHECK(is_entire(tan(make(1.5, 1.7))));  // straddles pi/2
  CHECK(is_entire(tan(make(0.0, 4.0))));  // spans more than a period
  CHECK(is_entire(tan(entire())));
  // A branch-interior interval stays bounded and increasing.
  Interval t = tan(make(0.1, 0.2));
  CHECK(!is_entire(t));
  CHECK(t.lo <= t.hi);
}

TEST_CASE("domains restrict toward empty") {
  CHECK(is_empty(log(make(-2.0, -1.0))));
  CHECK(is_empty(log2(make(-2.0, -1.0))));
  CHECK(is_empty(log1p(make(-3.0, -2.0))));
  CHECK(is_empty(asin(make(2.0, 3.0))));
  CHECK(is_empty(acos(make(-3.0, -2.0))));
  CHECK(is_empty(acosh(make(0.0, 0.5))));
  CHECK(is_empty(atanh(make(-3.0, -1.5))));
  // Partial-domain input is clipped, not rejected.
  CHECK(log(make(-1.0, 4.0)).lo == -kInfv);  // approaches 0+ -> -inf
  CHECK(equal(asin(make(-2.0, 0.0)), asin(make(-1.0, 0.0))));
  CHECK(acosh(make(0.5, 2.0)).lo >= 0.0);
}

TEST_CASE("unbounded inputs and limits") {
  CHECK(exp(make(-kInfv, 0.0)).lo == 0.0);
  CHECK(exp(make(-kInfv, 0.0)).hi <= 1.0 + 1e-15);
  CHECK(exp(make(0.0, kInfv)).hi == kInfv);
  CHECK(expm1(make(-kInfv, 0.0)).lo == -1.0);
  CHECK(log(make(0.0, kInfv)).lo == -kInfv);
  CHECK(atanh(make(0.5, 1.0)).hi == kInfv);
  CHECK(sinh(make(-kInfv, kInfv)).lo == -kInfv);
  CHECK(cosh(make(-kInfv, kInfv)).hi == kInfv);
}

TEST_CASE("empty propagates through every function") {
  const Interval e = empty();
  CHECK(is_empty(exp(e)));
  CHECK(is_empty(log(e)));
  CHECK(is_empty(sin(e)));
  CHECK(is_empty(cos(e)));
  CHECK(is_empty(tan(e)));
  CHECK(is_empty(sqrt(e)));
  CHECK(is_empty(atan2(e, point(1.0))));
  CHECK(is_empty(hypot(e, point(1.0))));
  CHECK(is_empty(pow(e, point(2.0))));
}

TEST_CASE("results are tight (about two ulps) at points") {
  auto narrow = [](const Interval& x) {
    return wid(x) <= 4.0 * std::ldexp(std::max(mag(x), 1.0), -52);
  };
  CHECK(narrow(exp(point(1.0))));
  CHECK(narrow(log(point(3.0))));
  CHECK(narrow(sin(point(1.0))));
  CHECK(narrow(cos(point(1.0))));
  CHECK(narrow(atan(point(0.5))));
  CHECK(narrow(cbrt(point(5.0))));
}

TEST_CASE("pow: integer exponents match pown; real powers via nonnegative base") {
  CHECK(equal(pow(make(-2.0, 3.0), point(2.0)), pown(make(-2.0, 3.0), 2)));
  CHECK(equal(pow(make(-2.0, 3.0), point(3.0)), pown(make(-2.0, 3.0), 3)));
  CHECK(holds(8.0, pow(point(2.0), point(3.0))));
  CHECK(holds(std::sqrt(2.0), pow(point(2.0), point(0.5))));
  // Negative base with non-integer exponent is undefined -> empty.
  CHECK(is_empty(pow(make(-3.0, -1.0), point(0.5))));
}

TEST_CASE("atan2 quadrants and branch cut") {
  CHECK(holds(std::atan2(1.0, 1.0), atan2(point(1.0), point(1.0))));    // ~ pi/4
  CHECK(holds(std::atan2(1.0, -1.0), atan2(point(1.0), point(-1.0))));  // ~ 3pi/4
  CHECK(holds(std::atan2(-1.0, -1.0), atan2(point(-1.0), point(-1.0))));
  // Box straddling the negative x-axis: conservative full-angle enclosure.
  Interval wrap = atan2(make(-1.0, 1.0), make(-2.0, -1.0));
  CHECK(wrap.lo <= -3.14);
  CHECK(wrap.hi >= 3.14);
}
